### PR TITLE
Search: fuzzy match dashboard names in frontend from full list

### PR DESCRIPTION
**Pull request by leeoniya**
works!

test locally with http://localhost:3000/?orgId=1&search=open&do-frontend-query

- intra-term fuzziness is set to `1` here, so will match `cart` and `coast` when searching `cat`. we can refine this further to a _total_ of `1` by implementing the `intraFilt` callback and checking matched term length against needle term length, or edit distance, etc.
- inter-term fuzziness is `Infinity` (by default), so will match multiple terms separated by arbitrary junk.
- out-of-order terms matching is done by permuting all terms in the needle (N!), so 2 terms == 2 searches, 3 terms == 6 searches, 4 terms == 24 searches, etc. the results are then combined, by skipping already-matched results; there are no fancy attempts to combine result stats to boost stuff that matches multiple permutations, or matches one permutation better than another.

the default sorting function prioritizes exact matches, prefix/suffix positions, density of terms, and match position. we can alter these priorities to a more typeahead-style (prioritizing match position and intra density) if needed.

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
Cool!! Does the bludge search do any fuzzy matching? 

I thought the frontend search stuff was removed when we moved backend search. Is the 
frontend searcher something we ever plan on using? Maybe we can use it for some of the list views we currently have frontend search for (where we can be farily sure the number of items is  low, like data source list) 

it has no tests as well but I assume that just because it was just a quick poc :)

**Comment by ryantxu** [comment]
> I thought the frontend search stuff was removed when we moved backend search.

It was... but we added hack to test if this path is plausible again.  The previous attempt was either too dumb, or too resource hungry to use for real.  This seems promising -- for sure with "small" numbers, though defining what we call small may be pretty flexible.  Performance wise it is fine with 10k items.

> Maybe we can use it for some of the list views we currently have frontend search for

for sure.  @leeoniya wrote this mainly for better support on a flame graph search, but we want to see how it works in general

> Does the bluge search do any fuzzy matching?

yes... and then we modified it to do less fuzzy matching ;)

File: public/app/features/search/service/frontend.test.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      query: 'ba',
+      kind: ['dashboard'],
+      location: 'General',
+      sort: 'name_sort',
+    };
+    const results = await frontendSearcher.search(query);
+
+    expect(results.view.fields.name.values.toArray()).toMatchInlineSnapshot(`
+      Array [
+        "bar dog",
```
Comment (ryantxu) [inline public/app/features/search/service/frontend.test.ts:50]: From the sample... is this sorted by `name_sort` or score?

I assume sorting by an explicit field is still on the TODO list and should likely throw an error

Comment (leeoniya) [reply public/app/features/search/service/frontend.test.ts:50]: yes it's sorted by score.

with the current settings the query can match substrings, so sorting by name is not a good default.

**Review by ryantxu — APPROVED** [review]

**Pull request merged**